identity-iota/src/tangle/message_version.rs

```diff
@@ -0,0 +1,41 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::Error;
+
+#[derive(Copy, Clone)]
+pub(crate) enum MessageVersion {
+  V1 = 1,
+}
+
+impl MessageVersion {
+  pub(crate) const CURRENT: Self = MessageVersion::V1;
+
+  /// Adds the current message version flag at the beginning of arbitrary data.
+  pub(crate) fn add_version_flag(mut data: Vec<u8>, message_version: MessageVersion) -> Vec<u8> {
+    let version_flag = message_version as u8;
+    data.splice(0..0, [version_flag].iter().cloned());
+    data
+  }
```

Comment:
Since both `add_version_flag` and `add_encoding_version_flag` both use `splice` it adds extra allocations unnecessarily. The methods can be combined and optimised by pre-allocating a vec with the data first, in `publish_json` or a new `pack_message` etc.

Reply:
I'm not sure where the unnecessary allocations happen, to my understanding, splice doesn't copy the entire vector in order to prepend items. And the idea of calling splice twice was for the sake of separation of concerns.

Reply:
> splice doesn't copy the entire vector in order to prepend items

Pre-pending with `splice` needs to allocate a new `Vec` with enough space and move the existing `Vec` into it as far as I know. It's efficient but this is still done twice which is unnecessary.

> And the idea of calling splice twice was for the sake of separation of concerns.

They're both operations that only happen when packing a message to be published, they can be combined.

Reply:
From https://doc.rust-lang.org/std/vec/struct.Vec.html#method.splice


> The input iterator replace_with is only consumed when the Splice value is dropped.
> 
> This is optimal if:
> 
> - The tail (elements in the vector after range) is empty,
> - or replace_with yields fewer or equal elements than range’s length
> - or the lower bound of its size_hint() is exact.
> 
> Otherwise, a temporary vector is allocated and the tail is moved twice.